### DISCUSSION TITLE
P10 sum

### DISCUSSION THREAD

**Discussion by dtswk** [BJReplay/ha-solcast-solar#51] [Ideas]
Firstly thank you for the awesome resource..!

We are off grid so I make a lot of automations to run non critical workloads and very often use the P10 as the worst case call if I am going to risk running them before our batteries hit x percent.

I'm really new to HA but had a crack at writing some templates to create sums and thought I would share.

Maybe you would consider building in?

Matt

  - name: "solcast_pv_forecast_p10_remaining_today"
    unique_id: "solcast_pv_forecast_p10_remaining_today"
    unit_of_measurement: "kWh"
    icon: mdi:solar-power
    state: >
      {% set data = state_attr('sensor.solcast_pv_forecast_forecast_today', 'detailedHourly') %}
      {% set result = data | map(attribute='period_start') | list %}
      {% set combined = data | map(attribute='pv_estimate10') | list %}
      {% set time_now = as_timestamp(now()) %}
      {% set ns = namespace(result=0) %}
            {% for item in data %}
            {% if as_timestamp(item.period_start) > as_timestamp(now()) %}                
              {% set ns.result = ns.result + item.pv_estimate10 | float %}   
            {% endif %}
            {% endfor %}
      {% set state = ns.result %}
      {{ state | round(1) }}
      
  - name: "solcast_pv_forecast_p10_forecast_today"
    unique_id: "solcast_pv_forecast_p10_forecast_today"
    unit_of_measurement: "kWh"
    icon: mdi:solar-power
    state: >      
      {% set data = state_attr('sensor.solcast_pv_forecast_forecast_today', 'detailedHourly') %}
      {% set result = data | map(attribute='period_start') | list %}
      {% set combined = data | map(attribute='pv_estimate10') | list %}
      {% set time_now = as_timestamp(now()) %}
      {% set ns = namespace(result=0) %}
            {% for item in data %}
              {% set ns.result = ns.result + item.pv_estimate10 | float %}   
            {% endfor %}
      {% set state = ns.result %}
      {{ state | round(1) }}

  - name: "solcast_pv_forecast_p10_tomorrow"
    unique_id: "solcast_pv_forecast_p10_tomorrow"
    unit_of_measurement: "kWh"
    icon: mdi:solar-power
    state: >      
      {% set data = state_attr('sensor.solcast_pv_forecast_forecast_tomorrow', 'detailedHourly') %}
      {% set result = data | map(attribute='period_start') | list %}
      {% set combined = data | map(attribute='pv_estimate10') | list %}
      {% set time_now = as_timestamp(now()) %}
      {% set ns = namespace(result=0) %}
            {% for item in data %}
              {% set ns.result = ns.result + item.pv_estimate10 | float %}   
            {% endfor %}
      {% set state = ns.result %}
      {{ state | round(1) }}

  - name: "solcast_pv_forecast_p10_day_3"
    unique_id: "solcast_pv_forecast_p10_day_3"
    unit_of_measurement: "kWh"
    icon: mdi:solar-power
    state: >      
      {% set data = state_attr('sensor.solcast_pv_forecast_forecast_day_3', 'detailedHourly') %}
      {% set result = data | map(attribute='period_start') | list %}
      {% set combined = data | map(attribute='pv_estimate10') | list %}
      {% set time_now = as_timestamp(now()) %}
      {% set ns = namespace(result=0) %}
            {% for item in data %}
              {% set ns.result = ns.result + item.pv_estimate10 | float %}   
            {% endfor %}
      {% set state = ns.result %}
      {{ state | round(1) }}

**Comment by gcoan** [comment]
I asked a similar thing a while ago and Ozzee helped me craft template sensors for pv_10 today and tomorrow which is all I wanted to see.

He didn't want to add these to the core integration, think he wanted to keep it cleaner as these were easy for someone to create themselves if they wanted to,  but it would be good if these were in the documentation to help others.

My template sensors were created through the Home Assistant UI rather than in configuration.yaml, and are quite a bit simpler:

```
{% set d = state_attr('sensor.solcast_pv_forecast_today', 'detailedHourly') %}
{{ (d | sum(attribute='pv_estimate10')) | round(2)  }}
```

I think it's doing the same?

**Comment by dtswk** [comment]
Thanks for sharing, that's a much nicer approach.  The only thing that still might be helpful is the remaining for today as I discard anything that has gone past the time.

**Comment by dtswk** [comment]
Updated code  :-)

  - name: "solcast_pv_forecast_p10_remaining_today"
    unique_id: "solcast_pv_forecast_p10_remaining_today"
    unit_of_measurement: "kWh"
    icon: mdi:solar-power
    state: >
      {% set data = state_attr('sensor.solcast_pv_forecast_forecast_today', 'detailedHourly') %}
      {% set result = data | map(attribute='period_start') | list %}
      {% set combined = data | map(attribute='pv_estimate10') | list %}
      {% set time_now = as_timestamp(now()) %}
        {% set ns = namespace(result=0) %}
          {% for item in data %}
            {% if as_timestamp(item.period_start) > as_timestamp(now()) %}
              {% set ns.result = ns.result + item.pv_estimate10 | float %}
            {% endif %}
          {% endfor %}
      {% set state = ns.result %}
      {{ state | round(1) }}
      
  - name: "solcast_pv_forecast_p10_forecast_today"
    unique_id: "solcast_pv_forecast_p10_forecast_today"
    unit_of_measurement: "kWh"
    icon: mdi:solar-power
    state: >      
      {% set data = state_attr('sensor.solcast_pv_forecast_forecast_today', 'detailedHourly') %}
      {{ data | map(attribute='pv_estimate10') | sum | round(1) }}

  - name: "solcast_pv_forecast_p10_tomorrow"
    unique_id: "solcast_pv_forecast_p10_tomorrow"
    unit_of_measurement: "kWh"
    icon: mdi:solar-power
    state: >      
      {% set data = state_attr('sensor.solcast_pv_forecast_forecast_tomorrow', 'detailedHourly') %}
      {{ data | map(attribute='pv_estimate10') | sum | round(1) }}

  - name: "solcast_pv_forecast_p10_day_3"
    unique_id: "solcast_pv_forecast_p10_day_3"
    unit_of_measurement: "kWh"
    icon: mdi:solar-power
    state: >      
      {% set data = state_attr('sensor.solcast_pv_forecast_forecast_day_3', 'detailedHourly') %}
      {{ data | map(attribute='pv_estimate10') | sum | round(1) }}

**Reply by gcoan** [reply]
Doh!

I was just creating a fork to paste this example code into and PR it back to bjreply, started writing it up, went to look at the pv_today sensors and found that there are Total10 and Total90 attributes on these sensors now so there's no need to iterate around as I had done!

The pv_remaining_today doesn't have this so still needs the iterator you had
![image](https://github.com/BJReplay/ha-solcast-solar/assets/142018870/880876c3-42f9-4c12-b9e5-4198c93a4af9)

**Reply by autoSteve** [reply]
> The pv_remaining_today doesn't have this so still needs the iterator you had

Sounds like a challenge to me @gcoan... You _know_ I read this stuff.

**Reply by dtswk** [reply]
Wow thank you..!!!  I updated to 4.0.35 and not seeing those new attributes.  I did call the update service which worked.  Any other hints?

**Reply by autoSteve** [reply]
(Not sure how Geoff has those attributes. I don't. @gcoan you lucky duck. An automation?)

The attributes for estimate/estimate10/estimate90 for pv_remaining_today? If so, they don't exist yet. I did say it was a challenge, but a challenge for the present, not the past.

Targeting 4.0.36, and fingers busily coding tonight.

Good on you for giving 4.0.35 a go. Zero issues for that one on my 'prod' rig. Enjoy the silence.

**Reply by dtswk** [reply]
Hey @autoSteve may the coding gods be with you..!   They still are beating me with a stick at the moment.. lol    HA is a learning curve compared with Node-Red  :-)

**Reply by autoSteve** [reply]
So say we all.

I'm doing the estimate/estimate10/estimate90 thing for remaining today, plus also the today/tomorrow/+3/+4 etc. Attributes will be called ~~estimate/estimate10/estimate90~~, not Total/Total10/Total90 (which reeks of a Geoff automation).

Nope. They'll be called pv_estimate/pv_estimate10/pv_estimate90.

**Reply by autoSteve** [reply]
The gods, they shine...

<img width="1137" alt="image" src="https://github.com/BJReplay/ha-solcast-solar/assets/37229860/6e71a114-bb91-4aa2-b896-145dec32fddc">

**Reply by autoSteve** [reply]
Crappy test at the moment, because it's dark here...

<img width="953" alt="image" src="https://github.com/BJReplay/ha-solcast-solar/assets/37229860/242a076a-5492-4972-8bd6-e4103989f142">

**Reply by gcoan** [reply]
> (Not sure how Geoff has those attributes. I don't. @gcoan you lucky duck. An automation?)
> 
> The attributes for estimate/estimate10/estimate90 for pv_remaining_today? If so, they don't exist yet. I did say it was a challenge, but a challenge for the present, not the past.
> 

Confused @autoSteve 

My pv forecast today has attributes of total today, total 10 and total 90 and similar remaining today:
![image](https://github.com/BJReplay/ha-solcast-solar/assets/142018870/c3dcd347-13c3-4808-9eef-2994a30e42bd)

Definitely not from an automation, I have a couple of template sensors for the pv10 today and tomorrow that I created from the detailed half hour forecasts. I assumed these are now created by solcast ?

Edit: just checked and they are on pv tomorrow and d3. Definitely not me

**Reply by autoSteve** [reply]
That's got me spooked 👻. In developer tools I did a browser search of everything for total90 and got absolutely nothin'.

**Reply by gcoan** [reply]
> That's got me spooked 👻. In developer tools I did a browser search of everything for total90 and got absolutely nothin'.

When I get back home I'll look at my code as well 🤷‍♂️

**Reply by gcoan** [reply]
> > That's got me spooked 👻. In developer tools I did a browser search of everything for total90 and got absolutely nothin'.
> 
> When I get back home I'll look at my code as well 🤷‍♂️

Mystery solved ...

solcast_pv_forecast_today/tomorrow/day3 etc doesn't have these extra attributes 

predbat_pv_today/tomorrow/d3 which is created by the Predbat automation system which manages my battery and inverters, does have these, it creates the attributes from the solcast forecast integration (or can poll solcast directly)

**Reply by autoSteve** [reply]
@dtswk, The branch https://github.com/BJReplay/ha-solcast-solar/tree/Post-4.0.35 has a commit of `coordinator.py` and `solcastapi.py` that add the breakdown attributes to forecast sensors. (And based on the latest update from Geoff these attributes will not collide with PredBat 🦇, just sit alongside them as a tiny additional amount of data.)

If you're willing, and you grok the Linux gods, then you could download these from the commit, and replace your two 4.0.35 files in `/homeassistant/custom_components/solcast_solar/`. Restart, and magic should happen. Hopefully. This is ⍺ code at the mo... But an extra tester would be ace, especially someone who is enjoying sunshine at the moment, unlike me sitting in a dark fridge downunder.

**Reply by autoSteve** [reply]
If it all goes to custard, @dtswk, just re-download 4.0.35 from BJReplay in HACS. I'm tipping it won't.

**Reply by dtswk** [reply]
OMG, flames, flames spung out of my Home Assistant Green...!!!!   hahhaha, no seriously worked pefectly.  Just doing a cross check from my own code, give me 10 mins

**Reply by dtswk** [reply]
Yep cross checked all the day and the numbers balance against mine.  :-) So either we both fooked up or it's all good  :-)

![image](https://github.com/BJReplay/ha-solcast-solar/assets/100445670/b2e66737-b511-40b4-8583-6e7aa7a0d89a)


Matt

**Reply by autoSteve** [reply]
The chances of us both screwing it up are remote. This is great feedback so thank you @dtswk.

I'll probably give it quite a few days before proposing a v4.0.36 to give v4.0.35 time to "settle". At eleven downloads in two hours, and total radio silence from Europe I'm tipping that one will be uneventful. That and having a discussion tag of #69 (🤭), which deserves to stay a while.

**Reply by autoSteve** [reply]
It's hard to know what attributes folks will want for forecast/forecast10/forecast90 from which sensors.

So I have added them _all_, with the exception of per-day forecast not broken down by site.

Targeting v4.0.36 once the dust settles from... Umm... Not much v4.0.35 complaining going on, and with 46 downloads so far.

The branch https://github.com/BJReplay/ha-solcast-solar/tree/Post-4.0.35 may have gone a little bit attribute mad, because there's heaps more there. Including @gcoan's per-site breakdown request. Yep. Even per-site forecast10... Grab the modified files of the branch if you want to be with me for the testing. `__init.py__`, `coordinator.py` & `solcastapi.py`.

I see no down-side to this. Folks choose what they pluck from attributes using a template sensor, and grin. There is some data duplication, like `{site}` forecast peak being the same number for me as `{site}_pv_forecast` forecast peak, but that is because the forecast option for me is the average, not one of the other options. If `forecast10` were the option set, then `{site}` forecast peak would be the `{site}_pv_forecast10` forecast peak value. If that makes sense... Keeping the duplication saves any automations/templates from being more complex.

There's also a debug log chatter-reduction in this, and heaps of logging modifications to make the non-debug logs less "SHOUT THE NAME SOLCAST!!!". Folks can read... The debug shouting makes sense, because I can just '/SOLCAST' and search easily in `less`.

So, @gcoan, your "and other sensors if it makes sense" in the FR wish is granted. 🧞

**Reply by gcoan** [reply]
> The branch https://github.com/BJReplay/ha-solcast-solar/tree/Post-4.0.35 may have gone a little bit attribute mad, because there's heaps more there. Including @gcoan's per-site breakdown request

You just don't want any more FR's for more attributes.  Do it once and it's done with forever?

Interesting different approach to life.  Oziee didn't want to add more sensors and attributes, taking the view that the data was there in the existing forecast attributes and so we could template it out if we wanted to.  But things change and we have a new 🧹 on the keyboard.

Anyway, more is always better isn't it, more money, more food, more time with the significant other?

**Reply by autoSteve** [reply]
More 🍺 for now and I'm a happy camper.

More FRs? Bring 'em. If there's gaps that aren't known about, then they should be considered and hopefully filled.

Oziee mightn't have had enough time on his hands. Who knows? Maybe he thought it would just result in more issues raised.

Yes, the data was kind of there to be had (for the uber-into-HA who could do it without breaking Google in the process), but my approach to automation is that it should be just that. Less config. Less Googling and wrangling YAML. And if an automation integration can make for less work, then it's doing its job... Isn't that the idea of automation? 🤷‍♂️

If this is a monumentally stupid idea, then please beat me over the head with why the attributes shouldn't be there to be had.

**Reply by autoSteve** [reply]
(Per-site forecast10 for the daily peak could _not_ have been templated...)

**Reply by gcoan** [reply]
> (Per-site forecast10 for the daily peak could _not_ have been templated...)

"Just" a simple matter of turning debug mode on, reading the debug messages from the last fetch into a string, parsing the string into a list, parsing the list to find the appropriate 10% values, summing them and multiplying by 1000 for kW.

Trivial exercise for the reader, got to give them something to do.... 😂

**Reply by autoSteve** [reply]
Bwahahahahaa. ROTFL. 🤣 (Except it's really cold here. Mental note to self: Get. Up. Off. Floor. Before. Freeze.... When... Summer?)

**Reply by autoSteve** [reply]
Your debug logging will be "quieter" with this next proposed update, BTW @gcoan. I suppressed @isorin's sensor updates because they were getting out of hand with the additional attribute requests (and he only included them to verify his wonderful performance optimisations). A flag now needs to be enabled at the top of `solcastapi.py` to reveal them. But do give this code an alpha test if you've got the time and inclination. It's running just fine on my 'prod' HA.

**Reply by isorin** [reply]
@autoSteve Yeah, thanks for doing this, I was thinking to suppress the debug output for sensors as well.

**Reply by autoSteve** [reply]
PR #71 will likely become pre-release v4.0.36 unless something happens with v4.0.35. It's in final review state with BJ, and there's a healthy number of v.35 downloads to not indicate issues with that one, nor issues reported for v.31 GA, which has over 2,000 downloads.

For now, Independence Day: Resurgence, because July 4th. 🎆

**Reply by autoSteve** [reply]
Released.

**Reply by gcoan** [reply]
> For now, Independence Day: Resurgence, because July 4th. 🎆

Over in Blighty all the talk is about the General Election that's also on today.  All the talk is of a landslide victory for the left/centrist opposition party after 14 years of the right/centrist (but increasingly extreme right wing) Conservative party.

We shall see.  Can hardly contain my excitement

**Reply by autoSteve** [reply]
<img width="239" alt="image" src="https://github.com/BJReplay/ha-solcast-solar/assets/37229860/8746931e-8d00-47ad-b3eb-4452114d2573">

**Comment by dtswk** [comment]
There you go  :-)   IMHO the P90 is generally less intresting as I think you would usually want to base on most likely which the standard forecast and maybe worst case if like me off grid so being paranoid.  The the map method would work for both and you could easily setup a bunch of sensors based on P90.  I "think" the loop is needed for the date comparison, I see there is a filter for map but haven't looked at it too closely.

**Comment by EddieSpaghettie** [comment]
Im confused : The above p10 calculation formula for me produces a number for remaining for today of :  5.7kw

If i set solcast to use : Forecast Estimate 10 (whuch i thought is the same) that creates a value of : 7.74kw

**Reply by dtswk** [reply]
Make sure both are remaining otherwise you are comparing the full day vs remaining

**Reply by autoSteve** [reply]
Or install 4.0.36 beta, recently released, @EddieSpaghettie, and use the pre-calculated values that are now attributes for all sensors. 😉

**Reply by EddieSpaghettie** [reply]
I am on 4.0.36beta and thats what im using. The pre-calculated is giving different to the formula above

**Comment by EddieSpaghettie** [comment]
This is what Im using:

{{ state_attr('sensor.solcast_pv_forecast_forecast_remaining_today', 'pv_estimate10') | float(0) }}
{{ states('sensor.solcast_pv_forecast_p10_remaining_today') | float(0) }}

Top Gives : 6.31now
Bottom Gives : 5.7

**Reply by EddieSpaghettie** [reply]
name: "solcast_pv_forecast_p10_remaining_today"
unique_id: "solcast_pv_forecast_p10_remaining_today"
unit_of_measurement: "kWh"
icon: mdi:solar-power
state: >
{% set data = state_attr('sensor.solcast_pv_forecast_forecast_today', 'detailedHourly') %}
{% set result = data | map(attribute='period_start') | list %}
{% set combined = data | map(attribute='pv_estimate10') | list %}
{% set time_now = as_timestamp(now()) %}
{% set ns = namespace(result=0) %}
{% for item in data %}
{% if as_timestamp(item.period_start) > as_timestamp(now()) %}
{% set ns.result = ns.result + item.pv_estimate10 | float %}
{% endif %}
{% endfor %}
{% set state = ns.result %}
{{ state | round(1) }}

**Reply by autoSteve** [reply]
Ripper feedback.

So which is correct?

The code uses the exact same 'math' to calculate remaining. It's the same function called, just with `forecast10` used from the values.

I'm genuinely interested should it be getting it wrong.

6.31kWh with some time passed since your last comment seems plausible. Does that match if you now switch to use forecast10 overall?

**Reply by autoSteve** [reply]
> `{% if as_timestamp(item.period_start) > as_timestamp(now()) %}`

Should that be `>` or `>=` ??? #notsure.

**Reply by EddieSpaghettie** [reply]
I got no idea which is correct but something is not the same. 
Currently one gives : 5.83
Other One : 3.1

**Reply by EddieSpaghettie** [reply]
I am using the setting :

![image](https://github.com/BJReplay/ha-solcast-solar/assets/58969883/c726598c-27a9-492f-8c14-de502ca5a2aa)

![image](https://github.com/BJReplay/ha-solcast-solar/assets/58969883/63261739-173b-4f19-9699-bfa0a3d3dc98)

{{ state_attr('sensor.solcast_pv_forecast_forecast_remaining_today', 'pv_estimate10') | float(0) }} gives 5.57

{{ states('sensor.solcast_pv_forecast_p10_remaining_today') | float(0) }} : Gives 3.1

**Reply by autoSteve** [reply]
Slow down. Take a breath.

Rig Solcast for `forecast10`. Now, go to Developer Tools, and filter by 'solcast', and find the remaining for rest of day. Look at the value. Compare the value to the attributes. it will match `pv_forecast10`. Now compare to your template sensor that is doing the loop. You will likely see a different value. Now...

I suspect that `{% if as_timestamp(item.period_start) > as_timestamp(now()) %}` should read `{% if as_timestamp(item.period_start) >= as_timestamp(now()) %}`. Change it. And quick reload.

Do the values from the template sensor doing the loop, and the attribute now match?

**Reply by EddieSpaghettie** [reply]
Ok but how do we know the formula above is correct or incorrect? Maybe the built in calculations on the current beta is incorrect?

**Reply by autoSteve** [reply]
Okay. 5.65 and 5.57 are close, but they should be equal. Use the Developer Tools method.

**Reply by EddieSpaghettie** [reply]
They are the same when i check but its late afternoon by me and sun is behind clouds so they changed before i could paste the images

**Reply by autoSteve** [reply]
Right, so what happens if you change your template sensor doing the loop to `{% if as_timestamp(item.period_start) >= as_timestamp(now()) %}` and do a quick reload? Does that match the sensor attribute?

**Reply by EddieSpaghettie** [reply]
I fixed the formula as suggested but its still lower

**Reply by EddieSpaghettie** [reply]
![image](https://github.com/BJReplay/ha-solcast-solar/assets/58969883/0de1b324-7b96-4dc2-8760-1e164401d541)

**Reply by EddieSpaghettie** [reply]
Can you perhaps post the formula from the code. Maybe there is something else different?

**Reply by autoSteve** [reply]
There is something off with the loop yaml is what I am thinking.

Like I said, the exact same code is calculating remaining today, just using a differing forecast value. If `_use_data_field` is None then it uses the currently selected estimate in `get_forecast_pv_estimates()`. Others can comment. I got nothin'

```
    def get_forecast_remaining_today(self, _use_data_field=None) -> float:
        """Return remaining forecasted production for today"""
        # time remaining today
        start_utc = self.get_now_utc()
        end_utc = self.get_day_start_utc() + timedelta(days=1)
        res = 0.5 * self.get_forecast_pv_estimates(start_utc, end_utc, site=None, _use_data_field=_use_data_field)
        return res

    def get_forecasts_remaining_today(self) -> Dict[str, Any]:
        res = {}
        for _data_field in ('pv_estimate', 'pv_estimate10', 'pv_estimate90'): res[_data_field] = self.get_forecast_remaining_today(_data_field)
        return res
...
(from get_forecast_pv_estimates()...)
            _data_field = self._use_data_field if _use_data_field is None else _use_data_field
```

**Reply by EddieSpaghettie** [reply]
Unless it has something to do with GMT Times? Im on GMT+2
Forecast Today/Tomorrow/Day After tomorrow are all the same. Is just remaining today thats not the same. Ill use the new built in one then rather. Thank you :)

**Reply by autoSteve** [reply]
GMT+10 here, so... It don't go weird here that I've noticed, being _really_ from the future in Oz, but I'll keep an eye on it, which will be easier than GMT+2. Thx for the feedback @EddieSpaghettie.

**Comment by EddieSpaghettie** [comment]
![image](https://github.com/BJReplay/ha-solcast-solar/assets/58969883/4358767a-2c68-490f-a084-5a74429a4eb4)

![image](https://github.com/BJReplay/ha-solcast-solar/assets/58969883/c62b403c-19ee-4432-bc2e-173d5f97aacb)

**Reply by autoSteve** [reply]
It fascinates me why `pv_estimate` is so close to `pv_estimate90`. That's optimistic. And within the imprecision of a `real` number when doing a calc, so essentially the same number. The numbers we get are the numbers we have to work with I guess...

`3.1` is horribly low to even the `estimate10` by comparison, so yeah, stare at the yaml for the issue.

**Reply by EddieSpaghettie** [reply]
Im in South Africa. So maybe the sun does some different things here haha. And its winter

**Reply by EddieSpaghettie** [reply]
I removed that yaml code completely. Will use the new built in ones you kindly provided in the latest beta!

**Reply by autoSteve** [reply]
(I _really_ should round those `real` numbers to four decimal places. The imprecision makes my eye twitch...)

#goescode...

**Reply by autoSteve** [reply]
Done. In 4.0.next...

**Reply by EddieSpaghettie** [reply]
Whoop whoop

**Reply by dtswk** [reply]
Interesting, was fine for me in same timezone. Time comparison didn't honour timezone then?
